### PR TITLE
Opera 0.5.0

### PR DESCRIPTION
**Pull request by tadeboro**
This monster of a PR contains second batch of changes that were needed to get opera into a functioning state.

### CODE REVIEW

File: src/opera/template/node.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    #
+    # TOSCA functions
+    #
+    def get_property(self, params):
+        host, prop, *rest = params
+
+        if host != "SELF":
+            raise DataError(
+                "Accessing non-local stuff is bad. Fix your service template."
```
Comment (matejart) [inline src/opera/template/node.py]:
```suggestion
                "Accessing non-local stuff is bad. Fix your service template and use SELF."
```

File: src/opera/template/relationship.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    #
+    # TOSCA functions
+    #
+    def get_property(self, params):
+        host, prop, *rest = params
+
+        if host != "SELF":
+            raise DataError(
+                "Accessing non-local stuff is bad. Fix your service template."
```
Comment (matejart) [inline src/opera/template/relationship.py]:
```suggestion
                "Accessing non-local stuff is bad. Fix your service template and use SELF."
```

File: tests/opera/parser/tosca/test_void.py [hunk 3]
```diff
@@ -4,13 +4,13 @@
 from opera.parser.utils.location import Location
 
 
-class TestBare:
-    @pytest.mark.parametrize("data", [1, 2.3, False, "string", None])
-    def test_bare_simple(self, data):
-        assert Void(data, Location("path", 2, 3)).bare == data
+class TestBuild:
+    pass
```
Comment (matejart) [inline tests/opera/parser/tosca/test_void.py]: Would it make sense to add a bunch of `#TODO`s here?

**Comment by matejart** [comment]
Oh, and I should say that I did try to understand a bit deeper, but in the time available that isn't a lot. I also test-ran the opera against the supplied examples.

File: examples/nginx_openstack/library/nginx/server/playbooks/reload.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  tasks:
+    - name: Restart nginx
+      service:
+        name: nginx
+        state: restarted
+
+    - name: Place temporary file
+      copy:
+        dest: /tmp/opera.mark
+        content: "{{ marker }}"
```
[the remark this answers is not in the document]
Comment (tadeboro) [reply examples/nginx_openstack/library/nginx/server/playbooks/reload.yaml]: This is actually a test if the intrinsic functions work on relationship operations ;) It has no other purpose actually and it should be removed once we have a proper set of integration tests.

File: src/opera/instance/relationship.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    #
+    # TOSCA functions
+    #
+    def get_attribute(self, params):
+        host, attr, *rest = params
+
+        if host not in ("SELF", "SOURCE", "TARGET"):
+            raise DataError(
+                "Accessing non-local stuff is bad. Fix your service template."
```
[the remark this answers is not in the document]
Comment (tadeboro) [reply src/opera/instance/relationship.py]: This local in the sense that we are not accessing other parts of the topology directly. We can still walk along the edges of the graph and access non-local values. I agree that the message could be a bit clearer but I would postpone the fix a bit. The error messages in this part of the code need a rather heavy refactoring and should start reporting at least some information about the error location.

File: src/opera/instance/relationship.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        if attr not in self.attributes:
+            raise DataError("Instance has no '{}' attribute".format(attr))
+        return self.attributes[attr].eval(self)
+
+    def get_property(self, params):
+        host, prop, *rest = params
+
+        if host not in ("SELF", "SOURCE", "TARGET"):
+            raise DataError(
+                "Accessing non-local stuff is bad. Fix your service template."
```
[the remark this answers is not in the document]
Comment (tadeboro) [reply src/opera/instance/relationship.py]: See the reply to the previous comment.

**Comment by tadeboro** [comment]
I fixed the issues that were discovered during the review. I would keep the bad error messages intact as an incentive to refactor the whole error reporting pipeline. Does this sound reasonable?

**Pull request merged**